### PR TITLE
bpo-32077: Documentation: Some Unicode object functions don't indicate whether they return a new reference

### PR DESCRIPTION
**Pull request by lioncash**
Recreates #4472 because I mistakenly thought it was merged a while back and removed my fork on Github, which ended up orphaning that pull request (sorry about that!)

Original pull request message:

--- 

Also adds missing functions. Makes the documentation more comprehensive in terms of indicating whether or not a function returns a new reference.

I've gone over the entries quite a few times, and I'm fairly confident I've amended most of the existing functions and added the new ones that weren't yet added. If I've happened to miss any, just say the word and I'll add them.

I've also noticed the API documentation for bytes objects is also missing entries, which I've made a local commit for, but wasn't sure if it was a good idea to intermix it in this change. I can PR it separately or include it in this one depending on whichever is preferable.

https://bugs.python.org/issue32077

### CODE REVIEW

File: Doc/data/refcounts.dat [hunk 1]
```diff
@@ -1636,35 +1647,39 @@ PyUnicode_FromEncodedObject:const char*:errors::
...
 PyUnicode_Decode:const char*:encoding::
 PyUnicode_Decode:const char*:errors::
 
 PyUnicode_DecodeUTF16Stateful:PyObject*::+1:
 PyUnicode_DecodeUTF16Stateful:const char*:s::
-PyUnicode_DecodeUTF16Stateful:int:size::
+PyUnicode_DecodeUTF16Stateful:Py_ssize_t:size::
 PyUnicode_DecodeUTF16Stateful:const char*:errors::
 PyUnicode_DecodeUTF16Stateful:int*:byteorder::
 PyUnicode_DecodeUTF16Stateful:int*:consumed::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat]:
```suggestion
 PyUnicode_DecodeUTF16Stateful:Py_ssize_t*:consumed::
```

Comment (lioncash) [reply Doc/data/refcounts.dat]: Ah, good catch, amended the commit

File: Doc/data/refcounts.dat [hunk 2]
```diff
@@ -1673,95 +1688,142 @@ PyUnicode_AsEncodedString:PyObject*:unicode::
...
+PyUnicode_EncodeUTF32:Py_ssize_t:size::
+PyUnicode_EncodeUTF32:const char*:errors::
+PyUnicode_EncodeUTF32:int:byteorder::
+
 PyUnicode_DecodeUnicodeEscape:PyObject*::+1:
 PyUnicode_DecodeUnicodeEscape:const char*:s::
-PyUnicode_DecodeUnicodeEscape:int:size::
+PyUnicode_DecodeUnicodeEscape:Py_ssize_t:size::
 PyUnicode_DecodeUnicodeEscape:const char*:errors::
+PyUnicode_DecodeUnicodeEscape:const char**:first_invalid_escape::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat]: Remove this.

File: Doc/data/refcounts.dat [hunk 3]
```diff
@@ -1673,95 +1688,142 @@ PyUnicode_AsEncodedString:PyObject*:unicode::
...
 
 PyUnicode_AsUnicodeEscapeString:PyObject*::+1:
 PyUnicode_AsUnicodeEscapeString:PyObject*:unicode::
 
 PyUnicode_DecodeRawUnicodeEscape:PyObject*::+1:
 PyUnicode_DecodeRawUnicodeEscape:const char*:s::
-PyUnicode_DecodeRawUnicodeEscape:int:size::
+PyUnicode_DecodeRawUnicodeEscape:Py_ssize_t:size::
 PyUnicode_DecodeRawUnicodeEscape:const char*:errors::
+PyUnicode_DecodeRawUnicodeEscape:const char**:first_invalid_escape::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat]: Remove this.

File: Doc/data/refcounts.dat [hunk 4]
```diff
@@ -1810,25 +1883,32 @@ PyUnicode_Join:PyObject*::+1:
...
 PyUnicode_Find:int:direction::
 
+PyUnicode_FindChar:Py_ssize_t:::
+PyUnicode_FindChar:PyObject*:str:0:
+PyUnicode_FindChar:Py_UCS4:ch::
+PyUnicode_FindChar:Py_ssize_t:start::
+PyUnicode_FindChar:Py_ssize_t:end::
+PyUnicode_FindChar:int:direction::
+
 PyUnicode_Count:int:::
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat]:
```suggestion
 PyUnicode_Count:Py_ssize_t:::
```

File: Doc/data/refcounts.dat [hunk 5]
```diff
@@ -1810,25 +1883,32 @@ PyUnicode_Join:PyObject*::+1:
...
 PyUnicode_Count:PyObject*:str:0:
 PyUnicode_Count:PyObject*:substr:0:
-PyUnicode_Count:int:start::
-PyUnicode_Count:int:end::
+PyUnicode_Count:Py_ssize_t:start::
+PyUnicode_Count:Py_ssize_t:end::
 
 PyUnicode_Replace:PyObject*::+1:
 PyUnicode_Replace:PyObject*:str:0:
```
Comment (serhiy-storchaka) [inline Doc/data/refcounts.dat:1912]: `maxcount` should be `Py_ssize_t`.

**Review by serhiy-storchaka — COMMENTED** [review]
It is better to for review to not squash commits.

**Comment by lioncash** [comment]
@serhiy-storchaka Ah, sorry about that, will keep addressed comments as a separate commit in the future.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**